### PR TITLE
SCRD-3476 Implemented deployment for adding compute nodes

### PR DESCRIPTION
**Pull request by ghost**
This checkin includes:
1. Added deployment processes for adding compute nodes
2. Fixed issuse when deployment is in progress, user switches to
the page which is loaded by installWizard.
3. Update global or persisted states to use the names which make
more sense.
4. When deployment is in progress, hide replace menu item in servers
summary, disable deploy button and  wipe disk option in add servers
5. Fixed the issue when all playbooks are completed, it relaunches playbooks
in playbook process.
6. Deal with various loading masks.
7. Handle playload per playbook when isUpdateMode is set
8. Fixed the styling issue in accordion when there are more than 1 compute roles
9. Added validation before launch playbooks for adding compute. Show error log
in a modal diaglog where there are validation errors.
10. When move to next page, clean up playbookStatus
11. Added errors handleing in the udpateServers page.
12. Fixed issuess regarding using string for boolean attributes in login page

Node:
1. This change  depends on the ardana-service changes
https://gerrit.suse.provo.cloud/#/c/4710
2. Need to deal with extraVars
3. Need to check duplicate ip-addr, mac-addr, ilo-ip when user drags
from auto/manual servers
4. Testd a happy path in dev environment, more testing and error handling
will come.
5. Need to handle activation

### CODE REVIEW

**Comment by jeremy-moffitt** [comment]
at a glance this seems to work ok, I still need to look at the code. One thing I've noticed is that the loading mask is used a *lot* and its not always clear whats loading... for example, after a new node has been added... or while the configprocessor is being kicked off. I wonder if theres some way we can alert the user to why the screen is blocked by the mask

**Comment by guglo** [comment]
loading mask is invoked when there is loading in the wizard, or a loading data for this page or or config processor kicks off...since those processes takes some time...maybe once user clicks 'Deploy' , add a warning confirmation modal?

**Comment by jeremy-moffitt** [comment]
> loading mask is invoked when there is loading in the wizard, or a loading data for this page or or config processor kicks off...since those processes takes some time...maybe once user clicks 'Deploy' , add a warning confirmation modal?

Hmmm will have to think about it a bit...we just need to make sure the user knows something is going on... our load mask doesnt have text on it, but I wonder if thats something we should look into

**Comment by guglo** [comment]
Talked to @jeremy-moffitt , will look into making the loading mask for validation more obvious with some text like validating...since it is a long loading

File: src/localization/bundles/en.json [hunk 1]
```diff
@@ -553,6 +553,20 @@
...
+    "server.addserver.compute.deploy": "Deploying Compute Servers in Progress",
+    "server.addserver.deploy.confirm": "The deployment for new servers will take a while. It will validate and commit data and run playbooks.  \n\n Are you sure you want to proceed? Select Yes to proceed.",
+    "server.deploy.progress.wipe-disks": "Wipe Disks",
+    "server.deploy.progress.gen-hosts-file": "Generate Host Files",
+    "server.deploy.progress.addserver.deploy": "Deploy New Servers",
+    "server.deploy.progress.update-monasca": "Update Monasca",
+    "server.deploy.addserver.complete": "Successfully Added Compute Servers",
+    "server.addserver.validate.error.title": "Failure of validating newly added servers",
     "server.addserver.error.get-deployed-servers": "Failed to get deployed servers. {0}",
     "server.addserver.wipedisk.message": "This will set to wipe disk for all newly added servers"
```
Comment (jeremy-moffitt) [inline src/localization/bundles/en.json]: not part of this commit, but this wording is not quite right

File: src/localization/bundles/en.json [hunk 2]
```diff
@@ -553,6 +553,20 @@
     "server.replace.prepare.failure": "Failed to prepare for replacing server",
     "update.commit.failure": "Failed to commit update changes. {0}",
     "install.progress.failure": "Failed to install SLES OS",
+    "server.retrieve.discovered.servers.error": "Failed to retrieve discovered servers. {0}",
+    "server.retrieve.manual.servers.error": "Failed to retrieve manually added servers. {0}",
+    "server.validating": "Validating Changes...",
+    "server.addserver.prepare": "Preparing Adding Servers in Progress",
+    "server.addserver.prepare.failure": "Failed to prepare for adding servers",
+    "server.addserver.deploy.failure": "Failed to deploy newly added servers",
+    "server.addserver.compute.deploy": "Deploying Compute Servers in Progress",
```
Comment (jeremy-moffitt) [inline src/localization/bundles/en.json]: I would remove "in Progress" from this string

**Review by jeremy-moffitt — CHANGES_REQUESTED** [review]
When I try to add a server, I get the load mask, and I can see the yes/no modal in the background, but never get a chance to click it (yes or not)
As soon as validation finishes, if it succeeds it goes into the first set of playbooks, which seem to be fine... then after clicking "next" I end up on a screen that says "Deploying Compute Servers in Progress" and is just stuck there with nothing going on

Note that if the validation fails, the yes/no modal is no longer applicable and shouldnt be shown

**Comment by jeremy-moffitt** [comment]
can confirm the confirmation dialog issue is fixed... tried again and did not see the hang problem on the 2nd page, will need to try a 3rd time to validate

**Comment by jeremy-moffitt** [comment]
I switched back to the mid-size model and am seeing the issue again where the 2nd page of the deployment is just blank... will need to investigate further before committing

**Comment by jeremy-moffitt** [comment]
per our conversation, it looks like the logic that determines whether a host is new is not working correctly... in some cases its deciding midway through that there are no new hosts, and stopping dead in its tracks... in other cases its deciding that the existing compute host is new and attempting to run the playbook on both that host *and* the newly added one.... also its confusing that the generated name is whats shown to the user at the end rather than the ID which they assigned to it while filling out the form

**Comment by guglo** [comment]
@jeremy-moffitt  thanks for the review and testing . The wired issues are caused by getting the stabled cached cloudmodel.yaml. Chatted with Gary and will get the non-cached data which seems to solve the problem.. Will do more testing and push fixed version shortly.

File: src/localization/bundles/en.json (hunk 2 above)
Comment (guglo) [reply src/localization/bundles/en.json]: ok will fix.

File: src/localization/bundles/en.json (hunk 1 above)
Comment (guglo) [reply src/localization/bundles/en.json]: can you be more specific? which one need fix?

**Comment by guglo** [comment]
The latest checkin depends https://gerrit.suse.provo.cloud//#/c/4776/ if want to try 

Test Notes: 
Here are the test data I used
server_id | ip_addr | mac_address | server-group | nic-mapping | role | Ipmi-addr
Myserver-mid | 10.243.140.30 | 5c:b9:01:88:d4:ce | RACK1 | HP-DL360-6PORT |   | 10.1.12.207
Myserver-mid2 | 10.243.140.35 | 5c:b9:01:88:d4:cc | RACK2 | HP-DL360-6PORT |   | 10.1.12.208
Myserver-mid3 | 10.243.140.36 | 5c:b9:01:88:d4:cd | RACK3 | HP-DL360-6PORT |   | 10.1.12.209
Myserver-kvm | 192.168.110.8 | d6:70:c1:36:43:ff | RACK1 | MY-2PORT-SERVER |   | 192.168.109.6
Myserver-kvm2 | 192.168.110.10 | d6:70:c1:36:43:fe | RACK1 | MY-2PORT-SERVER |   | 192.168.109.8

If in mid scale model, if set nic-mapping for example RACK3 for all 3 new nodes, it passes validation, but create some hosts without hostname and ardana-ansible-host. Since I have to use hostname to go with --limit, I will skip the host without hostname
if cloudmodel has record with empty hostname , need to have ardana-service fix https://gerrit.suse.provo.cloud//#/c/4776/

after deployment, newly added hosts will still have edit and delete action because of the mock deployed hostnames mock json doesn't include ardana-cp1-comp0004 and ardana-cp1-comp0005, ardana-cp1-comp0006
Also is use entry kvm ...after deployment, the newly added hosts will not have edit and delete action becasue ardana-cp1-comp0002 and ardana-cp1-comp0003 are in the mock json

**Comment by jeremy-moffitt** [comment]
I just tried this out and on the first page of the deployment, the "Pre-Deployment" playbook is green, but the others didnt start, do I need a different backend to test this?

**Comment by guglo** [comment]
There are some issues related playbook name changes. The newly updated version rebased with master and run with latest ardana-service master

**Comment by jeremy-moffitt** [comment]
the latest patch works for me, taking a look at the code next

File: src/pages/AddServers.js [hunk 3]
```diff
@@ -32,34 +34,90 @@ class AddServers extends BaseUpdateWizardPage {
...
+      // indicator of this loading
+      loading: true,
+      // show confirm dialog when user clicks Deploy
+      showDeployConfirmModal: false
     };
   }
 
   componentWillMount() {
-    fetchJson('/api/v1/clm/model/deployed_servers')
+    fetchJson('/api/v1/clm/model/deployed_servers', undefined, true, true)
```
Comment (jeremy-moffitt) [inline src/pages/AddServers.js]: it might be nice to know what the undefined is standing in for, but not an urgent fix

File: src/pages/AddServers.js [hunk 4]
```diff
@@ -32,34 +34,90 @@ class AddServers extends BaseUpdateWizardPage {
...
+    postJson('/api/v1/clm/config_processor')
+      .then(() => {
+        this.setState({validating: false});
+        let pages = this.assembleProcessPages();
+        let opProps = {'deployedServers': this.state.deployedServers};
+        this.props.startUpdateProcess('AddServer', pages, opProps);
+      })
+      .catch((error) => {
+        this.setState({validating: false});
+        this.setState({validationError: error.value? error.value.log : error.toString()});
```
Comment (jeremy-moffitt) [inline src/pages/AddServers.js]: maybe a space before the ?

File: src/pages/AddServers.js [hunk 5]
```diff
@@ -163,11 +257,13 @@ class AddServers extends BaseUpdateWizardPage {
         <div className='wizard-content unlimited-height'>
           {this.isValidToRenderServerContent() && this.renderAddPage()}
         </div>
-        {this.state.errors && this.renderErrorBanner()}
+        {this.renderDeployConfirmModal()}
+        {this.renderValidationErrorModal()}
+        {this.state.errorBanner && this.renderGetDeployedServerErrorBanner()}
```
Comment (jeremy-moffitt) [inline src/pages/AddServers.js]: "renderGetDeployedServerErrorBanner" is descriptive, if perhaps a bit long ;-)

File: src/pages/AddServers/CompleteAddServers.js [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+* See the License for the specific language governing permissions and
+* limitations under the License.
+**/
+
+import React from 'react';
+import BaseUpdateWizardPage from '../BaseUpdateWizardPage.js';
+import { LoadingMask } from '../../components/LoadingMask.js';
+import { translate } from '../../localization/localize.js';
+import { alphabetically } from '../../utils/Sort.js';
+
```
Comment (jeremy-moffitt) [inline src/pages/AddServers/CompleteAddServers.js]: theres no description or anything in this whole file that tells me what its used for

File: src/pages/AddServers/DeployAddServers.js [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+import { ErrorBanner, WarningMessage } from '../../components/Messages.js';
+import { PlaybookProgress } from '../../components/PlaybookProcess.js';
+import { translate } from '../../localization/localize.js';
+import {
+  STATUS, WIPE_DISKS_PLAYBOOK, ARDANA_GEN_HOSTS_FILE_PLAYBOOK,
+  SITE_PLAYBOOK, MONASCA_DEPLOY_PLAYBOOK
+} from '../../utils/constants.js';
+import { fetchJson } from '../../utils/RestUtils.js';
+
+
```
Comment (jeremy-moffitt) [inline src/pages/AddServers/DeployAddServers.js]: I think its fine to have these commented out ansible commands here, but it would be nice to have a one liner or two explaining what they are

File: src/pages/AddServers/DeployAddServers.js [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+    payload: {limit: {}}
+  },
+  {
+    name: MONASCA_DEPLOY_PLAYBOOK,
+    label: translate('server.deploy.progress.update-monasca'),
+    playbooks: [MONASCA_DEPLOY_PLAYBOOK + '.yml'],
+    payload: {tags: 'active_ping_checks'}
+  }
+];
+
```
Comment (jeremy-moffitt) [inline src/pages/AddServers/DeployAddServers.js]: the classes in this file have no description

File: src/pages/AddServers/DeployAddServers.js [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+    // this will take some time
+    // will request with no-cache
+    if(!this.props.operationProps.newHosts) {
+      this.setState({loading: true});
+      fetchJson(
+        '/api/v1/clm/model/cp_internal/CloudModel.yaml', undefined, true, true
+      )
+        .then((cloudModel) => {
+          let newHosts = this.getAddedComputeHosts(cloudModel);
+          let cleanedHosts = newHosts.filter(host => host['hostname'] !== undefined);
```
Comment (jeremy-moffitt) [inline src/pages/AddServers/DeployAddServers.js]: as we discussed this is a lot of extra logic to cover for a bug... no reason to remove it for now, maybe add a comment about the bugzilla ID of the bug?

File: src/pages/Login.js [hunk 10]
```diff
@@ -133,11 +133,11 @@ class LoginPage extends Component {
             <div className='content-right col-sm-4 col-sm-offset-1'>
               <div className='header'>{translate('login.header')}</div>
               <form onSubmit={this.handleLogin}>
-                <input type='text' className='rounded-corner' required='true' autoFocus='true'
+                <input type='text' className='rounded-corner' required={true} autoFocus={true}
```
Comment (jeremy-moffitt) [inline src/pages/Login.js]: this reads kinda weird, required and autoFocus are the reactJs interpretations of the boolean `true` ?

**Review by jeremy-moffitt — APPROVED** [review]
theres a fair amount of small clean-ups that could make this a lot easier for someone else to look at in the future, particularly around some of the class definitions. 

The code does work in my dev environment

File: src/pages/AddServers.js (hunk 4 above)
Comment (guglo) [reply src/pages/AddServers.js]: k will fix

File: src/pages/AddServers.js (hunk 3 above)
Comment (guglo) [reply src/pages/AddServers.js]: k  undefined is the init param for fetchJson...will add comments

File: src/pages/AddServers.js (hunk 5 above)
Comment (guglo) [reply src/pages/AddServers.js]: k, will fix

File: src/pages/AddServers/CompleteAddServers.js (hunk 6 above)
Comment (guglo) [reply src/pages/AddServers/CompleteAddServers.js]: It is used at line 42

File: src/pages/AddServers/DeployAddServers.js (hunk 7 above)
Comment (guglo) [reply src/pages/AddServers/DeployAddServers.js]: k , will remove them

File: src/pages/AddServers/DeployAddServers.js (hunk 8 above)
Comment (guglo) [reply src/pages/AddServers/DeployAddServers.js]: don't understand what you mean

File: src/pages/AddServers/DeployAddServers.js (hunk 9 above)
Comment (guglo) [reply src/pages/AddServers/DeployAddServers.js]: ok ..will file a bug for ansible and update with comments

File: src/pages/Login.js (hunk 10 above)
Comment (guglo) [reply src/pages/Login.js]: I fixed it because sometimes I got error/warning complains about that string...it expects boolean. updating the boolean stops the complains.

**Review by jeremy-moffitt — APPROVED** [review]
